File: app/src/main/java/com/jsoft/jcomic/praser/CartoonMadEpisodeParser.java
package com.jsoft.jcomic.praser;

import com.jsoft.jcomic.helper.EpisodeDTO;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class CartoonMadEpisodeParser extends EpisodeParser {
    public CartoonMadEpisodeParser(EpisodeDTO episode, EpisodeParserListener listener) {
        super(episode, listener, "BIG5");
    }
    protected void getEpisodeFromUrlResult(List<String> result) {
        episode.setImageUrl(new ArrayList<String>());
        String imageUrl_0 = "";
        String imageUrl_1 = "";
        String imageUrl_3 = "";
        int numPage = 0;
        for (String s: result) {
            Pattern p = Pattern.compile(".*<img src=\"(https?:\\/\\/[a-zA-Z0-9:\\.]+)?(.+)(\\d\\d\\d)(&.+=.+)?\" border=\"0\" oncontextmenu='return false' width=\"\\d+\">.*");
            //<img src="comicpic.asp?file=/1152/932/002" border="0" oncontextmenu='return false' width="700">
            //<img src="comicpic.asp?file=/1558/001/001&rimg=1" border="0" oncontextmenu='return false' width="700"></a>

                    Matcher m = p.matcher(s);
            if (m.matches()) {
                imageUrl_0 = m.group(1);
                imageUrl_1 = m.group(2);
                imageUrl_3 = m.group(4);
            }

            p=Pattern.compile(".*...<a class=pages href=(.+)>(\\d+)</a>.*");
            m = p.matcher(s);
            if (m.matches()) {
                numPage = Integer.parseInt(m.group(2));
            }

        }
        episode.setPageCount(numPage);

        List<String> imageUrlList = new ArrayList<String>();
        if (imageUrl_0 == null) {
            imageUrl_0 = "https://www.cartoonmad.com/m/comic/";
        }
        if (imageUrl_3 == null) {
            imageUrl_3 = "";
        }
        for (int k=1; k<= episode.getPageCount(); k++) {
            String imageUrl = imageUrl_0 + imageUrl_1 + String.format("%03d", k) + imageUrl_3;
            imageUrlList.add(imageUrl);
        }
        episode.setImageUrl(imageUrlList);
    }
}

File: app/src/main/java/com/jsoft/jcomic/praser/CartoonMadEpisodeParser.kt
package com.jsoft.jcomic.praser

import com.jsoft.jcomic.helper.EpisodeDTO
import com.jsoft.jcomic.helper.Utils
import java.net.URL
import java.util.*
import java.util.regex.Pattern

class CartoonMadEpisodeParser(episode: EpisodeDTO, listener: EpisodeParserListener) : EpisodeParser(episode, listener, "BIG5") {

    override fun getURLResponse(url: URL, encoding: String): ArrayList<String> {
        val path = if (url.path.startsWith("/m")) url.path.substring(2) else url.path
        return Utils.getURLResponse(URL("http://" + url.host + path), null, encoding)
    }

    override fun getEpisodeFromUrlResult(result: List<String>) {
        episode.imageUrl = ArrayList()

        //http://www.cartoonmad.com/m/comic/102900123095001.html
        var p = Pattern.compile(".*(\\d{4})\\d(\\d{3})\\d(\\d{3})(\\d{3}).*")
        var m = p.matcher(episode.episodeUrl)
        var bookId = ""
        var episodeId = ""
        if (m.matches()) {
            episode.pageCount = Integer.parseInt(m.group(3))
            bookId = m.group(1)
            episodeId = m.group(2)
        }
        //http://www.cartoonmad.com/m/comic/comicpic.asp?file=/1029/032/004&rimg=1
        val imageUrlList = ArrayList<String>()
        for (k in 1..episode.pageCount) {
            val imageUrl = "http://www.cartoonmad.com/comic/comicpic.asp?file=/" + bookId + "/" + episodeId + "/" + String.format("%03d", k) + "&rimg=1"
            //Log.d("jComics", imageUrl)
            imageUrlList.add(imageUrl)
        }
        episode.imageUrl = imageUrlList

    }
}
